### ISSUE TITLE
Add Entity search to Unified Dashboard Header

### ISSUE THREAD

**Issue by tofumatt** [google/site-kit-wp#4049]
## Feature Description

The "Entity Search" should be added to the Unified Dashboard Header, implemented in #4048.

Not expanded:

<img width="683" alt="Screenshot 2021-09-15 at 20 27 57" src="https://user-images.githubusercontent.com/90871/133497085-260b28f3-33f0-429a-aacd-14db381efbb8.png">

Expanded:

<img width="669" alt="Screenshot 2021-09-15 at 20 42 50" src="https://user-images.githubusercontent.com/90871/133500671-f0af54d7-d355-4a82-8a4c-8c870c04e392.png">

Expanded with search results:

<img width="689" alt="Screenshot 2021-09-15 at 20 42 56" src="https://user-images.githubusercontent.com/90871/133500688-b5b163a2-f708-48c8-b503-81bfd23d9ef6.png">

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* These features should all be implemented behind the `unifiedDashboard` feature flag.
    * The Entity Search should normally appear "not expanded" (see above). This should be covered by #4048.
    * The Entity Search _"URL Search"/magnifying glass icon button_ should show the search input when clicked—this is also covered in #4048, but highlighting it here.
* A new component for the Entity Search Input can be created in `assets/js/components/EntitySearchInput`.
* The Entity Search Input implemented in the header should behave similarly to the existing `URLSearchWidget`'s `PostSearcherAutoSuggest`—that component can likely serve as the basis for this component. Key difference between that component and this new one are:
    * Styling as per the Figma files for both desktop and mobile views:
        * The desktop Figma file is here: https://www.figma.com/file/kKYqiUL7Fj3yAF8HHBRUhz/Custom-Dashboard?node-id=1944%3A17087 (under the "Single Search: Full frames" section).
        * The mobile versions are available here: https://www.figma.com/file/VILRieNSLg2DOVyEgFBkXe/%5BMVP%5D-Generic-Dashboard?node-id=382%3A1699
        * There are no dedicated tablet styles, so the mobile styles should be used for tablet sizes, essentially until all header content can fit alongside the input (as seen in the Desktop styles).
    * A "Loading" state should be added to this component featuring preview blocks, instead of the current behaviour when shows "No results" when first searching, and then later shows the _previous results_ while searching. It's clear that whatever the past results were are being shown in the search results; they should be cleared when the input changes in this new widget and "No results" should only be shown once the result is complete and empty. Essentially, where it [currently says "No results" in this screencast](https://user-images.githubusercontent.com/90871/135259311-077e7aef-d5dd-4472-8543-3af1fad8218a.gif), should be preview/loading lines instead.
    * There should be no "View data" submit button like in the `URLSearchWidget`. Instead, clicking on an item in the search results should [navigate](https://github.com/google/site-kit-wp/blob/7ee265439d32c19415370d75eeca72f6505e252d/assets/js/googlesitekit/widgets/components/URLSearchWidget.js#L48) to the Entity Dashboard immediately.
        * Before this navigation occurs, the GA event `open_urldetails` (currently also exists in the widget) should be triggered. The event category for this new event should be "${viewContext}_headerbar" (using the current view context for the `viewContext` placeholder). The code to create this event would be:

```js
const viewContext = useContext( ViewContextContext );

trackEvent( `${ viewContext }_headerbar`, 'open_urldetails' );
```

## Implementation Brief
* Using `assets/js/components/PostSearcherAutoSuggest.js`,
	* Add 3 new props:
		* `placeholder`: empty string by default, used to display placeholder text for the input.
		* `setIsLoading`: optional function. This should be provided from the parent component to set a loading flag.
		* `onKeyDown`: optional function for keyboard events for the input.
	* In the `useEffect` hook which calls the API passing on `debouncedValue`, call `setIsLoading` if defined, with `true`. When we have a response, then set `setIsLoading` to `false`.
	* Using `useEffect`, check if `searchTerm` is empty to clear the results, by setting `setResults` to `[]`. This is to clear previous search results when the user starts typing a new term.
	* Add `placeholder` and `onKeyDown` props to the `ComboboxInput`, and calling the namesake props if they are defined.
	* Using the `useFeature` hook, passing `unifiedDashboard` as parameter, display `ComboboxOption` rendering `noResultsMessage` only if they flag is not enabled. This means that current functionality will remain the same, but for the unified dashboard, it will not be displayed. Instead the loading state will be displayed in the parent component.
* Create `assets/js/components/EntitySearchInput.js` which exports the `EntitySearchInput` functional component.
	* `EntitySearchInput` uses most of the code from `assets/js/googlesitekit/widgets/components/URLSearchWidget.js` with the following changes:
		* Remove the `Widget` component, `Title or URL` label and all `mdc-layout-*` markup.
		* Primarily the `PostSearcherAutoSuggest` and `Button` will be used.
		* The class name for the main parent container should be `googlesitekit-entity-search` instead of `googlesitekit-post-searcher`.
	* Add 2 new state variables:
		* `isOpen`: `boolean` with `false` as default value.
		* `isLoading`: `boolean` with `false` as default value.
	* Using the `useEffect` hook, with `match` as dependency, check if there's a permalink to navigate the user to the details URL and track the necessary event. It's the same logic used in the `onClick` function.
	* For the tracking, refer to the AC for the new event category to be used.
	* Update the `onClick` callback function to toggle `isOpen`.
	* Update the `Button` component used:
		* Add the `trailingIcon` prop to render the appropriate icon using the following logic:
			* If `isOpen` is `true`, display the `close` icon. Refer to the mobile designs in Figma which contains the close icon, compared to the desktop one where the incorrect icon is displayed when the search is in the expanded state.
			* If `isOpen` is `false`, then display the search icon.
			* The icon SVGs should be placed in `assets/svg/`.
		* The label should be `URL Search`. The latter should be visible only when `isOpen` is `true`, meaning when the input is displayed, only the icon is displayed and the label is hidden.
	* Update `PostSearcherAutoSuggest` to:
		* add the `placeholder` prop, passing `Enter title or URL…` as value.
		* set the `setCanSubmit` to an anonymous function which does nothing.
		* add the `setIsLoading` prop, passing the function to change the `isLoading` state variable.
		* update the `id` prop to `entity-urlsearch-autocomplete`.
		* add the `onKeyDown` prop which the callback function will check if the `ESCAPE` key (imported from `@wordpress/keycodes`) has been pressed and `isOpen` is `true` to set `isOpen` to `false`.
	* `PostSearcherAutoSuggest` should be visible only when `isLoading` is `false`.
	* If `isLoading` is `true`, then render the `ProgressBar` component, with `small` and `compress` as props within the `googlesitekit-entity-search` container.
	* Using the `useRef` hook, attach a ref to the `googlesitekit-entity-search` container.
	* Using the `useEffect` hook, listen when `isOpen` changes. If `isOpen` is `true`, set the focus to the input text, else set the focus to the `button`. This means when the user clicks on button to display the input text, the focus is already set on the input for the user to start typing.
	* The `googlesitekit-entity-search--is-open` class name is added to the parent container rendering the markup when `isOpen` is `true`.
* Create `assets/js/components/EntitySearchInput.stories.js` which contains the story for `EntitySearchInput`.
* Create `assets/sass/components/entity-search/_googlesitekit-entity-search.scss` which will contain the styles for `EntitySearchInput` as per the designs in Figma.
	* When the search is in expanded state, i.e `isOpen` is `true` and thus having the `googlesitekit-entity-search--is-open` class name, set the button to use `position: absolute` so that the button is now displayed within the input.
	* The `ProgressBar` component should be absolutely positioned within the input text, below the close icon, having the same width as the icon. Refer to this comment for an example: https://github.com/google/site-kit-wp/issues/4049#issuecomment-935725206

**UPDATE**: when working on the new entity search component, add the event tracking defined in  for the `URLSearchWidget` component.

### Test Coverage
* No new tests to be added.

## QA Brief

* The Entity Search component in the Unified Dashboard should look and behave as per AC/IB.
* Should work on all screen sizes.
* Some design elements are incorrect in the Screenshot/Figma, see discussion below.
    * Desktop View shows `search` icon instead of `close` icon when search is expended/open. 
    * See comment for the loaders' appearance hint (not available on Figma). 
* Check for Keyboard navigation that was added in the Code Review Phase. 
    * Press `ESC` to close the expended search.
    * Press `Up` / `Down` to to select result from the list. Should have visual feedback. 
        * Pressing `Enter` should navigate to the selected entity page. 
    * Pressing `Enter` when the search term is a (case insensitive) match should result in navigating to the entity page. If it's not ma match, pressing `Enter` should do nothing. 
* The appropriate tracking even should be sent to GA on navigating to an entity page. (Use Google Analytics Debugger to verify this).

## Changelog entry

* Add Entity search to Unified Dashboard Header

**Comment by tofumatt** [comment]
The GA Events for the URL Search in #4057 should fire here as well, when the URL Search item is clicked and the navigation is triggered.

**Comment by tofumatt** [comment]
@felixarntz I think these ACs are good but there's one thing I wanted to confirm with you: the GA event we send for the Entity Search navigation. In the old Search Widget on the dashboard we use `trackEvent( `${ viewContext }_urlsearch-widget`, 'open_urldetails' );`, but that seems wrong here. I've suggested `trackEvent( `${ viewContext }_urlsearch-header`, 'open_urldetails' );` (eg. changing the category to reflect the new event).

Is that okay? If so I think this can be moved to IB.

**Comment by felixarntz** [comment]
@tofumatt The ACs look good. For the event we need to use `${ viewContext }_headerbar`, as that is the same category the other header bar items use, and something we care about here specifically is what users interact with the most. I've updated that in the ACs, so this can be moved forward.

One other thing I just came across: Do we already have an issue to introduce new view context constants for the "main dashboard" and "entity dashboard"? We should probably be using those instead of the one for "dashboard" and "page dashboard" whenever the feature flag is active. If not, let's create those.

**Comment by asvinb** [comment]
@tofumatt
> A "Loading" state should be added to this component featuring preview blocks, instead of the current behaviour when shows "No results" 

Displaying the `PreviewBlock` component beneath the input text looks a bit weird when I tested. How about a more subtle animation which is displayed where the magnifying/close icon is displayed?
![image](https://user-images.githubusercontent.com/31818/135865067-c8a8a5bc-20fb-4af4-9a40-c788d0339c91.png)

**Comment by ivankruchkoff** [comment]
Take a look at IB for #4048 as it also covers `assets/js/components/EntitySearchInput.js`

**Comment by tofumatt** [comment]
A note that we might want to adjust the IB in #4048 based on this IB, just making a note here for me to review it and make sure these don't overlap too much (and that they align).

**Comment by asvinb** [comment]
@tofumatt If we are to put the indicator next to the close button, it should be positioned within the input text similar to the close icon which led me asking whether if we add the indicator next to it, it will just take some unnecessary space within the text input. So I thought about 2 possible solutions:

1. ![image](https://user-images.githubusercontent.com/31818/136165558-a08e6e27-20d0-41c2-8b43-0deaf2486d26.png) The loading indicator is positioned below the icon, thus we don't use extra space.
2. Secondly I tested using our `PreviewTable` component. For whatever reason, earlier I was testing with `PreviewBlock` (most likely since preview block was mentioned in the AC 😅 ) and it does not look bad. Here is a screenshot of how it'll look: ![image](https://user-images.githubusercontent.com/31818/136165858-8f14ffb3-8165-4701-ab7e-769bc66363f8.png)

So let me know what you think. I kinda prefer the subtle loading indicator but if it's too small am afraid it can be missed, especially it'll be in a corner, whereas the other solution using `PreviewTable` is much more visible.

Let me know what you think and I'll update the IB.

**Comment by tofumatt** [comment]
I definitely prefer the subtle loading indicator (option 1), especially for mobile screens where the autocomplete will really dominate the whole screen. I think it's visible enough and it's fine that it's subtle. Let's go with that option where it's below the icon, I think that's really slick. 👍🏻

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@kuasha420 @tofumatt I've found a few observations from testing:

1) On desktop (using Chrome and Safari) the styling of the results dropdown is broken when you have a notification 

E.g. `wp-admin/admin.php?page=googlesitekit-dashboard&slug=pagespeed-insights&notification=authentication_success#googlesitekit-pagespeed-header`

![image](https://user-images.githubusercontent.com/73545194/140938862-d4050d9f-51ef-418c-a460-0e4f9b4fb9c5.png)

2) The hover colour on the search button looks lighter than other menu items, i.e. date picker

![image](https://user-images.githubusercontent.com/73545194/140939096-d87f98fd-a3ba-4e63-ba4c-1ad82500bcae.png)

3) When I type a search term that does not have a URL for Post or page, I do not see any reference to ‘no results’ 

In the AC it says: 
> "No results" should only be shown once the result is complete and empty.

![image](https://user-images.githubusercontent.com/73545194/140939691-edb48d6d-dfc5-4550-90ba-a1609559c690.png)

4) From a UI/UX perspective, I do not find the loading state to be in a good placement, at the far end of the input field, near the close X icon.  I feel it needs to b e more prominent within the search box. Thoughts?

**Comment by tofumatt** [comment]
On very wide desktop browsers I could see that loading indicator being quite subtle, but it's: more instructive/visible than the previous UI that _had no loading indicator_ and works well on mobile screens I think.

We didn't have any UX to go on for that loading indicator so I think that can be a follow-up issue—could you file it separately? We might want some design feedback on that because I don't know what best practices for Material are here. But the rest of the points are worth addressing here I think. 👍🏻

**Comment by wpdarren** [comment]
@felixarntz @aaemnnosttv 

Following on from @tofumatt [comment](https://github.com/google/site-kit-wp/issues/4049#issuecomment-964210654), I do not know if this warrants a ticket, or, if I'm nit picking.  

Before I create a ticket, I would appreciate your thoughts. When searching it wasn't obvious that there's a loading indicator as I was typing in the search term. I agree with @tofumatt that it looks fine on mobile, but on desktop it feels it needs to be more prominent.  Is this something we should ticket, or happy as is?

**Comment by felixarntz** [comment]
@wpdarren Creating a new issue to review the loading indicator placement sounds good to me. Let's focus on your QA feedback points 1.-3. in this issue here as @tofumatt mentioned.

**Comment by kuasha420** [comment]
Submitted a [PR](https://github.com/google/site-kit-wp/pull/4331) fixing 2 and 3. Number 1 is a regression involving z-index and stacking context after the subHeader PR got merged. @asvinb Can you please have a look at this?

**Comment by asvinb** [comment]
Followup PR for fixing the z-index issue: https://github.com/google/site-kit-wp/pull/4334

**Comment by eugene-manuilov** [comment]
@wpdarren this is ready for QA.

**Comment by wpdarren** [comment]
### QA Update:  ✅ 

**Verified:**

- The Entity Search appears not expanded.   - [Screenshot](https://user-images.githubusercontent.com/73545194/141485586-8054276c-f8a3-4bf3-a037-37de38d546c6.png)
- The Entity Search URL Search/magnifying glass icon button shows the search input when clicked. - [Screenshsot](https://user-images.githubusercontent.com/73545194/141485746-eedac3fd-8ff8-4564-baa5-a20fab875e29.png)
- The Entity Search Input is based on the Styling as per the Figma files for both desktop and mobile views  - [Screenshot](https://user-images.githubusercontent.com/73545194/141486118-6277e97a-4502-4600-b7a4-3baadbaf37d0.png)
- A "Loading" state appears, instead of the current behaviour when shows "No results" when first searching
- A "No results" message appears only once the result is complete.  [Screenshot](https://user-images.githubusercontent.com/73545194/141486815-39c6c394-52b2-4596-8227-94be8dd4d5c3.png)
- When clicking on an item in the search results the user is navigated to the Entity Dashboard immediately.
- The GA event `open_urldetails` is triggered with an event category of `dashboard_headerbar` or `pagedashboard_headerbar` if using the search on the page dashboard. - [Screenshot](https://user-images.githubusercontent.com/73545194/141488565-7db0c27e-7b1a-4e10-8642-1908236e3aa6.png)
- The entity search works on all screen sizes.
- The Keyboard navigation that was added in the Code Review Phase functions as per QAB.
